### ISSUE TITLE
[API] Product create_product critical bug

### ISSUE THREAD

**Issue by Nils-Fredrik** [woothemes/woocommerce#8550]
Hi,

If you pass in the following to the ```create_product```method of the ```WC_API_Products```class (https://github.com/woothemes/woocommerce/blob/master/includes/api/class-wc-api-products.php#L205): 
```[product] => Array
        (
            [sku] => abc
        )```

```$id = wp_insert_post( $new_product, true );``` on line 265 (https://github.com/woothemes/woocommerce/blob/master/includes/api/class-wc-api-products.php#L265) will return a wp_error and an exception will be thrown. 

The exception is caught in line 293 (https://github.com/woothemes/woocommerce/blob/master/includes/api/class-wc-api-products.php#L293) which looks like this: 

```
			// Remove the product when fails
			$this->clear_product( $id );
			return new WP_Error( $e->getErrorCode(), $e->getMessage(), array( 'status' => $e->getCode() ) );
```

$id is still a wp_error object. So $this->clear_product() is called with a wp_error object as parameter. 

And that is where the real catastrophe occurs!

```
protected function clear_product( $product_id ) {
		if ( 0 >= $product_id ) {
			return;
		}
		// Delete product attachments
		$attachments = get_children( array(
			'post_parent' => $product_id,
			'post_status' => 'any',
			'post_type'   => 'attachment',
		) );
		foreach ( (array) $attachments as $attachment ) {
			wp_delete_attachment( $attachment->ID, true );
		}
		// Delete product
		wp_delete_post( $product_id, true );
	}
```

It passes the first test (```if ( 0 >= $product_id ) {```) with a warning and then it moves on and deletes absolutely every attachment in the database!

get_children() called with no post_parent returns all attachments. 

I propose two fixes for this: 
1) the clear_product() method: 
```
protected function clear_product( $product_id ) {
		if ( ctype_digit( $product_id ) && 0 >= $product_id ) {
			return;
		}

		// Delete product attachments
		$attachments = get_children( array(
			'post_parent' => $product_id,
			'post_status' => 'any',
			'post_type'   => 'attachment',
		) );

		foreach ( (array) $attachments as $attachment ) {
			wp_delete_attachment( $attachment->ID, true );
		}

		// Delete product
		wp_delete_post( $product_id, true );
	}
```
and 2) in the create_product method: 
``` 
} catch ( WC_API_Exception $e ) {
			// Remove the product when fails

		    if( ctype_digit( $id ) )    {
    			$this->clear_product( $id );
		    } 

			return new WP_Error( $e->getErrorCode(), $e->getMessage(), array( 'status' => $e->getCode() ) );
		}
``` 

I will suggest a fix tomorrow. 


Nils-Fredik

**Comment by claudiosmweb** [comment]
I'll check it

**Comment by Nils-Fredrik** [comment]
I have created a pull request with a fix to master. I think it also should be fixed in a 2.3.x release.

**Comment by Nils-Fredrik** [comment]
@claudiosmweb what do I do to fix this in a 2.3.x release?
